llvm/lib/Target/AMDGPU/GCNPreRAOptimizations.cpp

```diff
@@ -225,6 +254,712 @@ bool GCNPreRAOptimizationsImpl::processReg(Register Reg) {
+  uint16_t UnpackedOpcode = mapToUnpackedOpcode(I);
+
+  if (UnpackedOpcode == std::numeric_limits<uint16_t>::max())
+    return;
+  // Unpacked instructions
+  MachineInstrBuilder LoMul_MI =
+      BuildMI(MBB, I, DL, TII->get(UnpackedOpcode), LoMul);
+
+  if (Src0_Mods & SISrcMods::NEG)
+    Lo_src0_mods |= SISrcMods::NEG;
+
+  LoMul_MI.addImm(Lo_src0_mods);            // src0_modifiers
+  LoMul_MI.addReg(Src0_Lo, RegState::Kill); // src0
+
+  if (Src1_Mods & SISrcMods::NEG)
+    Lo_src1_mods |= SISrcMods::NEG;
+
+  LoMul_MI.addImm(Lo_src1_mods);            // src1_modifiers
+  LoMul_MI.addReg(Src1_Lo, RegState::Kill); // src1
+  LoMul_MI.addImm(ClampVal);                // clamp
+  // packed instructions do not support output modifiers. safe to assign them 0
+  // for this use case
+  LoMul_MI.addImm(0); // omod
+
+  // unpacked instruction with VOP3 encoding for Hi bits
+  unsigned Hi_src0_mods = 0;
+  unsigned Hi_src1_mods = 0;
+
+  MachineInstrBuilder HiMul_MI =
+      BuildMI(MBB, I, DL, TII->get(UnpackedOpcode), HiMul);
+  if (Src0_Mods & SISrcMods::NEG_HI)
+    Hi_src0_mods |= SISrcMods::NEG_HI;
+
+  HiMul_MI.addImm(Hi_src0_mods); // src0_modifiers
+  HiMul_MI.addReg(Src0_Hi,
+                  RegState::Kill); // select higher 16 bits if op_sel_hi is set
+
+  if (Src1_Mods & SISrcMods::NEG_HI)
+    Hi_src1_mods |= SISrcMods::NEG_HI;
+
+  HiMul_MI.addImm(Hi_src1_mods); // src0_modifiers
+  HiMul_MI.addReg(
+      Src1_Hi,
+      RegState::Kill); // select higher 16 bits from src1 if op_sel_hi is set
+  HiMul_MI.addImm(ClampVal); // clamp
+  // packed instructions do not support output modifiers. safe to assign them 0
+  // for this use case
+  HiMul_MI.addImm(0); // omod
+
+  // Shift HiMul left by 16
+  Register HiMulShifted = MRI.createVirtualRegister(RC);
+  MachineInstrBuilder HiMulShifted_MI =
+      BuildMI(MBB, I, DL, TII->get(AMDGPU::V_LSHLREV_B32_e64), HiMulShifted)
+          .addReg(ShiftAmt)
+          .addReg(HiMul);
+
+  SubRegID = 0;
+  if (DstMO.getSubReg())
+    SubRegID = DstMO.getSubReg();
+  // OR LoMul | (HiMul << 16)
+  MachineInstrBuilder RewriteBackToDst_MI =
+      BuildMI(MBB, I, DL, TII->get(AMDGPU::V_OR_B32_e64));
+  if (SubRegID) {
+    if (DstMO.isUndef()) {
+      RewriteBackToDst_MI.addDef(DstReg, RegState::Undef, SubRegID);
+    } else {
+      RewriteBackToDst_MI.addDef(DstReg, 0, SubRegID);
+    }
+  } else {
+    if (DstMO.isUndef()) {
+      RewriteBackToDst_MI.addDef(DstReg, RegState::Undef);
+    } else {
+      RewriteBackToDst_MI.addDef(DstReg);
+    }
+  }
+  RewriteBackToDst_MI.addReg(LoMul);
+  RewriteBackToDst_MI.addReg(HiMulShifted);
+
+  LIS->InsertMachineInstrInMaps(*LoMul_MI);
+  LIS->InsertMachineInstrInMaps(*HiMul_MI);
+  LIS->InsertMachineInstrInMaps(*HiMulShifted_MI);
+  LIS->InsertMachineInstrInMaps(*RewriteBackToDst_MI);
+  LIS->RemoveMachineInstrFromMaps(I);
+  I.eraseFromParent();
+  LIS->removeInterval(DstReg);
+  LIS->createAndComputeVirtRegInterval(DstReg);
```

Comment:
Are you also [re]computing the virtRegInterval for the use regs -- these may shift slightly due to SlotIndex numbering. Or, in the case where we are creating new regs to hold temporary results for 16 bit unpack, we'll need LiveIntervals for all those.